packages/server/src/__tests__/plugin/usageReporting/stats.test.ts

```diff
@@ -473,3 +478,97 @@ describe('Check type stats', () => {
     expect(contextualizedStats).toMatchSnapshot();
   });
 });
+
+describe('Add trace to report', () => {
+  const defaultHeader = new ReportHeader({
+    hostname: 'hostname',
+    agentVersion: `@apollo/server`,
+    runtimeVersion: `node latest`,
+    uname: 'uname',
+    executableSchemaId: 'schema',
+    graphRef: 'graph',
+  });
+  const referencedFieldsByType = Object.create(null);
+  referencedFieldsByType['type'] = new ReferencedFieldsForType({
+    fieldNames: ['field1', 'field2'],
+    isInterface: false,
+  });
+
+  it('add as stats if asTrace is false', () => {
+    const report = new OurReport(defaultHeader);
+    report.addTrace({
+      statsReportKey: 'key',
+      trace: baseTrace,
+      asTrace: false,
+      referencedFieldsByType,
+    });
+
+    expect(report.tracesPerQuery['key']?.trace?.length).toBe(0);
+    expect(
+      Object.keys(report.tracesPerQuery['key']?.statsWithContext?.map).length,
+    ).toBe(1);
+  });
+
+  it('add as stats if asTrace is true but trace is too large', () => {
+    const report = new OurReport(defaultHeader);
+    report.addTrace({
+      statsReportKey: 'key',
+      trace: baseTrace,
+      asTrace: true,
+      referencedFieldsByType,
+      maxTraceBytes: 10,
+    });
+
+    expect(report.tracesPerQuery['key']?.trace?.length).toBe(0);
+    expect(
+      Object.keys(report.tracesPerQuery['key']?.statsWithContext?.map).length,
+    ).toBe(1);
+  });
+
+  it('add as trace if asTrace is true and trace is not too large', () => {
+    const report = new OurReport(defaultHeader);
+    report.addTrace({
+      statsReportKey: 'key',
+      trace: baseTrace,
+      asTrace: true,
+      referencedFieldsByType,
+      maxTraceBytes: 500 * 1024,
+    });
+
+    expect(report.tracesPerQuery['key']?.trace?.length).toBe(1);
+    expect(
+      Object.keys(report.tracesPerQuery['key']?.statsWithContext?.map).length,
+    ).toBe(0);
+  });
+
+  it('add as trace if asTrace is true and trace is not too large and max trace size is left as default', () => {
+    const report = new OurReport(defaultHeader);
+    report.addTrace({
+      statsReportKey: 'key',
+      trace: baseTrace,
+      asTrace: true,
+      referencedFieldsByType,
+    });
+
+    expect(report.tracesPerQuery['key']?.trace?.length).toBe(1);
+    expect(
+      Object.keys(report.tracesPerQuery['key']?.statsWithContext?.map).length,
+    ).toBe(0);
+  });
+
+  it('add as trace if asTrace is true and max trace size is invalid', () => {
+    const report = new OurReport(defaultHeader);
+    report.addTrace({
+      statsReportKey: 'key',
+      trace: baseTrace,
+      asTrace: true,
+      referencedFieldsByType,
+      maxTraceBytes: 'one' as unknown as number,
```

Comment:
FWIW I don't really feel like we need a test of what happens if we call a purely internal function with an argument that doesn't pass TypeScript type-checking. (I also don't think the `isNan` check in the code itself adds much.)

Reply:
Fair enough. My thought process was just that I should be careful to maintain backwards-compatible functionality in case there was somehow any usages of the function that were somehow passing something unexpected in. But now that we're ignoring traces over 10mb on the ingress side it doesn't really matter.